### PR TITLE
Migrate family of `raw` client commands to GRPC V2.

### PR DESCRIPTION
**Pull request by vikt0r0**
## Purpose

Migrate the `raw` family of commands to use the GRPC V2 API; this is part of migrating the client to the GRPC V2 API feature.

Depends on https://github.com/Concordium/concordium-base/pull/317

## Changes
Implement helpers for gluing V2 API and `raw` commands of the client:

### Tasks
- Wrap output of `fromProto` in `Either` rather than `Maybe`. 
- Migrate `raw` commands.
- Misc. fixes and better comments in GRPC V2 `FromProto` instances.

### Client interface changes
- Remove `raw` client commands `startBaker`, `stopBaker`, `joinNetwork`, `leaveNetwork`.
- Remove node ID input from `BanNode` and `UnbanNode`.
- Add dump file path and `raw` flag to `DumpStart`.
- Change output of `GetBannedPeers` to display JSON instead of using leveraging `show`, e.g.

```
Right {peer_type: "Node" peers { node_id { value: "*" } ip { value: "10.10.10.10" } catchup_status: PENDING } peers { node_id { value: "*" } ip { value: "192.168.0.1" } catchup_status: PENDING } peers { node_id { value: "*" } ip { value: "192.168.0.100" } catchup_status: PENDING } peers { node_id { value: "*" } ip { value: "192.168.1.1" } catchup_status: PENDING } peers { node_id { value: "*" } ip { value: "192.168.1.100" } catchup_status: PENDING }}
```

to:

```
[
    "10.10.10.10",
    "192.168.0.1",
    "192.168.0.100",
    "192.168.1.1",
    "192.168.1.100"
]
```

- Change output of `GetPeerUptime` to display a `Word64` instead of `Maybe Word64`.
- Output of `GetTransactionStatusInBlock` when no transaction summary is present in response.

## Checklist
- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Comment by vikt0r0** [comment]
I used FIXME's here again to indicate the places where I am in doubt about how to approach things. I will make sure to go through all of these and make sure these are addressed before merging it into the feature branch.

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [inline src/Concordium/Client/Runner.hs:3823]: I'm not sure if I should be using `enqueue` here. Should I just be adding it to the list which represents the queue, and should the list be sorted by the effective update time?

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3715]: Use `logFatal` to get better error reporting.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3813]: Use `logFatal` instead of fail.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3466]: In light of the fact that the the node is missing an endpoint, please do the following

- split this command into the 4 ones
- GetBlockEvents
- GetBlockSpecialEvents
- GetChainParameters
- GetBlockPendingUpdates

That'll also be faster to use. Then we are not blocked by the node changes.

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3823]: Thanks for the fast review. So no reporting of invariant errors of the received stream is not sorted according to the effective update time?

File: src/Concordium/Client/GRPC2.hs
Comment (vikt0r0) [reply src/Concordium/Client/GRPC2.hs:806]: Yes, definitely. I'll address it.

File: src/Concordium/Client/GRPC2.hs
Comment (vikt0r0) [reply src/Concordium/Client/GRPC2.hs:1452]: I added that it is due to malformed payload data.

File: src/Concordium/Client/GRPC2.hs
Comment (vikt0r0) [reply src/Concordium/Client/GRPC2.hs:1582]: You are right. I changed all the instances of that error message to "convert" rather than "decode". Hope that's OK.

**Review by Bargsteen — COMMENTED** [review]
Nice work, had some minor comments to add to Ales' ;)

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3466]: Should `GetBlockPendingUpdates` output a JSON representation of the raw response or the response converted into a `PendingUpdates` instance containing the individual queues (lifting the helper functions I already implemented)?

I guess the former since it's a `raw` command?

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3481]: I'll close this conversation since we are splitting this function. Let me know if something else should be done.

File: src/Concordium/Client/GRPC2.hs
Comment (vikt0r0) [reply src/Concordium/Client/GRPC2.hs:256]: There was, but it's a leftover! Thanks for pointing it out! :-)

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3967]: I am not sure I understand, do you mean something like

`NodeActive (BakerConsensusInfo _ (PassiveBaker AddedButNotActiveInCommittee)) -> "No (registered as baker but not in currect epoch)."?`

Should it be the same for `getBakerCommitteeMember` above?

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:4143]: I believe you mentioned that conversion related error messages should probably not not be visible to the user when we discussed this at some point. What do you think is better here?

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:4151]: Yes, I will stop using ticks for naming things from now on. Would something like `getResponseValueId` be better?

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:3466]: Whatever is easier, really.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:4143]: I agree it should not be visible. But if there is a bug it is better that it is as clear as it can be. 

So propagating the internal error will be useful in a bug report.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:4151]: I think this function should be named `getResponseValue` and the one that maps it should be called something like `extractFromResponse`.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:3967]: I meant that it is OK to just return True/False here.

The elaborate cases was meant to list all the cases in the match statement. Although this part is debatable so if it is too verbose just keep what is there now with a catch-all case.

File: src/Concordium/Client/Runner.hs
Comment (Bargsteen) [reply src/Concordium/Client/Runner.hs]: Seems like you don't use `printResponseValueAsJSON` directly. So perhaps that is the solution.
But otherwise, I would have renamed the non-prime version to something with `map`.

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3682]: There are a lot of `fail`s scattered throughout Runner. Should they be changed to `logFatal`?

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs]: Concordium.Client.GRPC2 will replace Concordium.Client.GRPC post migration and it is a heavily referenced module in this file. Qualifying it will result in a lot of prefixes, so I think it may make the code more terse. `Concordium.Types` is already qualified, so that import should definitely be removed. I'll qualify the latter :-) thanks.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:3682]: Some of them probably should, but this is not to be done here.

That should be a separate PR. Some of them need to be revised to handle the error better, so we need to go through them one by one and see what should be done, see, e.g., #47

File: src/Concordium/Client/GRPC2.hs
Comment (vikt0r0) [reply src/Concordium/Client/GRPC2.hs:2786]: This was indeed not the case anyways :-)

File: src/Concordium/Client/GRPC2.hs
Comment (vikt0r0) [reply src/Concordium/Client/GRPC2.hs]: That's super neat. I used it in runner now, but I get the following error when adding it here for some reason:

```
concordium-base  >     Could not load module ‘Data.String.Interpolate’
concordium-base  >     It is a member of the hidden package ‘string-interpolate-0.3.1.2’.
concordium-base  >     Perhaps you need to add ‘string-interpolate’ to the build-depends in your .cabal file.
concordium-base  >     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
```

File: ChangeLog.md
Comment (vikt0r0) [inline ChangeLog.md:20]: Should this mention the unit? Before it printed a `Maybe Word64`, hence the change.

File: ChangeLog.md
Comment (vikt0r0) [inline ChangeLog.md:22]: This is due to some fields being elaborated a bit. For instance the baker ID is included when the node is part of a baker or finalization committee.

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [inline src/Concordium/Client/Runner.hs:4012]: I will rename this.

File: ChangeLog.md
Comment (Bargsteen) [reply ChangeLog.md:20]: Yes, in case someone relies on the output (a simple runner script fx).

File: ChangeLog.md
Comment (Bargsteen) [inline ChangeLog.md:14]: Something is slightly off in this sentence. Perhaps just?
```suggestion
    - `raw BanNode` and `raw UnbanNode` no longer support node IDs, now they take just an IP address.
```

File: ChangeLog.md
Comment (Bargsteen) [inline ChangeLog.md:19]:
```suggestion
    - Output of `raw GetBannedPeers` prints a JSON list containing strings, each corresponding to a banned IP.
```

File: ChangeLog.md
Comment (Bargsteen) [reply ChangeLog.md:22]: The fact that you made a comment here, suggests that you should add more details to this point in the changelog ;)

File: src/Concordium/Client/Runner.hs
Comment (Bargsteen) [inline src/Concordium/Client/Runner.hs:3657]: `logError` prefixes `Error:` on its own, so `FAIL:` seems redundant.

**Review by Bargsteen — APPROVED** [review]
LGTM apart from a few minor comments 👍

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:3657]: I agree.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3671]:
```suggestion
    -- |If the provided value is @Nothing@, a default value specified provided
    -- as the first parameter is returned. If the provided value is @Just t@ then t
    -- is parsed as a block hash. If this fails 'logFatal' is called to report and terminate the process.
```

The main point I'm trying to get across is that **there is nothing gained by mentioning "monad" here**. There is nothing from the concept of Monad that is relevant for this function.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3683]: Please be explicit that the `bool` parameter means "recurse" to parent.

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3831]:
```suggestion
                  "In current finalizer committee with baker ID " <> show bId <> "'."
```

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3816]:
```suggestion
                  "In current baker committee with baker ID '" <> show bId <> "'."
```

File: src/Concordium/Client/Runner.hs
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:4012]: Please do before merging :)

File: src/Concordium/Client/GRPC2.hs
Comment (abizjak) [inline src/Concordium/Client/GRPC2.hs:1730]: I don't think this is accurate a message.

**Review by abizjak — APPROVED** [review]

**Comment by abizjak** [comment]
Please handle all the comments before merging though :)

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3671]: Good point. This feedback is useful for me when writing comments in the future.

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3657]: If we want to retain the interface as much as possible should I perhaps just use `printStrLn` as before here then?

Comment (abizjak) [reply src/Concordium/Client/Runner.hs:3657]: It is not about retaining the interface here. Here the message is just ugly with Error: FAIL. Error indicates failure already.

Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3657]: Right, but doesn't the change from:

`putStrLn $ "FAIL: " <> x`
 
 to

`logError [[i|whatever: #{x}|]]`
 
change the output and hence the interface? I was just wondering if we should do as before and do `putStrLn` or use `logFatal` in this case?

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:4012]: Yes. I changed this in a later PR which I will request review for today. I hope that is alright.

File: src/Concordium/Client/Runner.hs
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3657]: I'll use `putStrLn` here for now to preserve the output. Let me know if it should be changed.

**Pull request merged**
